Failed tests (10679): TestSplitSnapshotRace_SplitWins fails (& leaks goroutines)

The following test appears to have failed:

[#10679](https://circleci.com/gh/cockroachdb/cockroach/10679):

```
W0106 14:45:11.633646 983 storage/store.go:1374  conflicting intents on [/Meta2/Max]: resolved? true
E0106 14:45:11.645222 983 storage/replica.go:1206  error executing raft command: unexpected value: raw_bytes:"\013\0333O\003\010\013\022\001m\032\002\377\377\"\006\010\004\020\004\030\001\"\006\010\002\020\002\030\002\"\006\010\003\020\003\030\003\"\006\010\005\020\005\030\004(\005" timestamp:<wall_time:3000000003 logical:54 > 
W0106 14:45:11.647263 983 kv/dist_sender.go:636  failed to invoke BeginTransaction [/Local/Range/RangeDescriptor/"m",/Min), ConditionalPut [/Local/Range/RangeDescriptor/"m",/Min), Put [/Meta2/Max,/Min): unexpected value: raw_bytes:"\013\0333O\003\010\013\022\001m\032\002\377\377\"\006\010\004\020\004\030\001\"\006\010\002\020\002\030\002\"\006\010\003\020\003\030\003\"\006\010\005\020\005\030\004(\005" timestamp:<wall_time:3000000003 logical:54 > 
I0106 14:45:11.647895 983 client/db.go:478  failed batch: unexpected value: raw_bytes:"\013\0333O\003\010\013\022\001m\032\002\377\377\"\006\010\004\020\004\030\001\"\006\010\002\020\002\030\002\"\006\010\003\020\003\030\003\"\006\010\005\020\005\030\004(\005" timestamp:<wall_time:3000000003 logical:54 > 
E0106 14:45:11.649718 983 storage/replica.go:1206  error executing raft command: unexpected value: raw_bytes:"\013\0333O\003\010\013\022\001m\032\002\377\377\"\006\010\004\020\004\030\001\"\006\010\002\020\002\030\002\"\006\010\003\020\003\030\003\"\006\010\005\020\005\030\004(\005" timestamp:<wall_time:3000000003 logical:54 > 
--- FAIL: TestSplitSnapshotRace_SplitWins (0.79s)
	client_test.go:495: storage/replica_command.go:1722: change replicas of 11 failed: unexpected value: raw_bytes:"\013\0333O\003\010\013\022\001m\032\002\377\377\"\006\010\004\020\004\030\001\"\006\010\002\020\002\030\002\"\006\010\003\020\003\030\003\"\006\010\005\020\005\030\004(\005" timestamp:<wall_time:3000000003 logical:54 > 
=== RUN   TestSplitSnapshotRace_SnapshotWins
W0106 14:45:11.678412 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:11.679299 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:11.778472 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:11.779474 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
--- FAIL: TestSplitSnapshotRace_SnapshotWins (0.20s)
	client_split_test.go:803: TODO(bdarnell): https://github.com/cockroachdb/cockroach/issues/3121
	<autogenerated>:31: Test appears to have leaked a store:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc82023cdc0)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:214 +0xf1a
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:201 +0x2e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820204230, 0xc8200cc1f0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
--
			/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:36 +0x34
		main.main()
			github.com/cockroachdb/cockroach/storage/_test/_testmain.go:498 +0x20a
=== RUN   TestStoreSplitReadRace
==================
WARNING: DATA RACE
Write by goroutine 89:
  github.com/cockroachdb/cockroach/storage_test.TestStoreSplitReadRace()
      /go/src/github.com/cockroachdb/cockroach/storage/client_split_test.go:867 +0x322
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc

Previous read by goroutine 52:
  github.com/cockroachdb/cockroach/storage.(*Replica).executeCmd()
      /go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:57 +0x282
  github.com/cockroachdb/cockroach/storage.(*Replica).executeBatch()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1449 +0x39c
  github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommandInBatch()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1320 +0x53c
  github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1231 +0x4b6
  github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1199 +0x314
--
  github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1609 +0x425
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x5f

Goroutine 89 (running) created at:
  testing.RunTests()
      /tmp/workdir/go/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /tmp/workdir/go/src/testing/testing.go:494 +0xe4
  github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck()
      /go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x2e
  github.com/cockroachdb/cockroach/storage_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/storage/main_test.go:36 +0x33
  main.main()
      github.com/cockroachdb/cockroach/storage/_test/_testmain.go:498 +0x209

Goroutine 52 (running) created at:
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x6f
  github.com/cockroachdb/cockroach/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1640 +0xb6
  github.com/cockroachdb/cockroach/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:545 +0xeae
  github.com/cockroachdb/cockroach/storage_test.(*multiTestContext).addStore()
      /go/src/github.com/cockroachdb/cockroach/storage/client_test.go:409 +0xc63
  github.com/cockroachdb/cockroach/storage_test.(*multiTestContext).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/client_test.go:226 +0x1702
--
I0106 14:45:11.904559 983 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
W0106 14:45:11.978453 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:11.978672 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:12.097014 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:12.097639 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
--- FAIL: TestStoreSplitReadRace (0.31s)
	<autogenerated>:31: Test appears to have leaked a store:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc82023cdc0)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:214 +0xf1a
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:201 +0x2e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820204230, 0xc8200cc1f0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
		
--
I0106 14:45:12.404901 983 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
W0106 14:45:12.478374 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:12.478835 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:12.578357 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:12.578572 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
--- FAIL: TestLeaderAfterSplit (0.44s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc82023cdc0)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:214 +0xf1a
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:201 +0x2e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820204230, 0xc8200cc1f0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
		
--
W0106 14:45:13.778341 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:13.778868 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:13.895335 983 storage/replica.go:1156  group 2 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
W0106 14:45:13.896367 983 storage/replica.go:1156  group 11 on store 3 failed to send message to 4: storage/raft_transport.go:163: unknown peer 4
--- PASS: Example_rebalancing (1.31s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	28.409s
=== RUN   TestBatchBasics
I0106 14:44:52.374122 1006 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0106 14:44:52.378614 1006 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I0106 14:44:52.385234 1006 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_event_test.go
+++ b/storage/client_event_test.go
@@ -224,7 +224,7 @@ func TestMultiStoreEventFeed(t *testing.T) {
 
 	// Replicate the default range.
 	rangeID := roachpb.RangeID(1)
-	mtc.replicateRange(rangeID, 0, 1, 2)
+	mtc.replicateRange(rangeID, 1, 2)
 
 	// Add some data in a transaction
 	pErr := mtc.db.Txn(func(txn *client.Txn) *roachpb.Error {

--- a/storage/client_merge_test.go
+++ b/storage/client_merge_test.go
@@ -343,9 +343,9 @@ func TestStoreRangeMergeNonCollocated(t *testing.T) {
 
 	// Replicate the ranges to different sets of stores. Ranges A and C
 	// are collocated, but B is different.
-	mtc.replicateRange(rangeA.RangeID, 0, 1, 2)
-	mtc.replicateRange(rangeB.RangeID, 0, 1, 3)
-	mtc.replicateRange(rangeC.RangeID, 0, 1, 2)
+	mtc.replicateRange(rangeA.RangeID, 1, 2)
+	mtc.replicateRange(rangeB.RangeID, 1, 3)
+	mtc.replicateRange(rangeC.RangeID, 1, 2)
 
 	// Attempt to merge.
 	rangeADesc = rangeA.Desc()

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -605,13 +605,13 @@ func TestUnreplicateFirstRange(t *testing.T) {
 
 	rangeID := roachpb.RangeID(1)
 	// Replicate the range to store 1.
-	mtc.replicateRange(rangeID, 0, 1)
+	mtc.replicateRange(rangeID, 1)
 	// Unreplicate the from from store 0.
-	mtc.unreplicateRange(rangeID, 1, 0)
+	mtc.unreplicateRange(rangeID, 0)
 	// Replicate the range to store 2. The first range is no longer available on
 	// store 1, and this command will fail if that situation is not properly
 	// supported.
-	mtc.replicateRange(rangeID, 1, 2)
+	mtc.replicateRange(rangeID, 2)
 }
 
 // TestStoreRangeDownReplicate verifies that the replication queue will notice
@@ -629,7 +629,7 @@ func TestStoreRangeDownReplicate(t *testing.T) {
 	rightKey := roachpb.Key("z")
 	{
 		replica := store0.LookupReplica(roachpb.RKeyMin, nil)
-		mtc.replicateRange(replica.RangeID, 0, 1, 2)
+		mtc.replicateRange(replica.RangeID, 1, 2)
 		desc := replica.Desc()
 		splitArgs := adminSplitArgs(splitKey, splitKey)
 		if _, err := replica.AdminSplit(splitArgs, desc); err != nil {
@@ -639,7 +639,7 @@ func TestStoreRangeDownReplicate(t *testing.T) {
 	// Replicate the new range to all five stores.
 	replica := store0.LookupReplica(keys.Addr(rightKey), nil)
 	desc := replica.Desc()
-	mtc.replicateRange(desc.RangeID, 0, 3, 4)
+	mtc.replicateRange(desc.RangeID, 3, 4)
 
 	// Initialize the gossip network.
 	var wg sync.WaitGroup
@@ -758,7 +758,7 @@ func TestProgressWithDownNode(t *testing.T) {
 	defer mtc.Stop()
 
 	rangeID := roachpb.RangeID(1)
-	mtc.replicateRange(rangeID, 0, 1, 2)
+	mtc.replicateRange(rangeID, 1, 2)
 
 	incArgs := incrementArgs([]byte("a"), 5)
 	if _, err := client.SendWrapped(rg1(mtc.stores[0]), nil, &incArgs); err != nil {
@@ -810,7 +810,7 @@ func TestReplicateAddAndRemove(t *testing.T) {
 
 		// Replicate the initial range to three of the four nodes.
 		rangeID := roachpb.RangeID(1)
-		mtc.replicateRange(rangeID, 0, 3, 1)
+		mtc.replicateRange(rangeID, 3, 1)
 
 		incArgs := incrementArgs([]byte("a"), 5)
 		if _, err := client.SendWrapped(rg1(mtc.stores[0]), nil, &incArgs); err != nil {
@@ -840,11 +840,11 @@ func TestReplicateAddAndRemove(t *testing.T) {
 		// Stop a store and replace it.
 		mtc.stopStore(1)
 		if addFirst {
-			mtc.replicateRange(rangeID, 0, 2)
-			mtc.unreplicateRange(rangeID, 0, 1)
+			mtc.replicateRange(rangeID, 2)
+			mtc.unreplicateRange(rangeID, 1)
 		} else {
-			mtc.unreplicateRange(rangeID, 0, 1)
-			mtc.replicateRange(rangeID, 0, 2)
+			mtc.unreplicateRange(rangeID, 1)
+			mtc.replicateRange(rangeID, 2)
 		}
 		verify([]int64{5, 5, 5, 5})
 
@@ -892,7 +892,7 @@ func TestRaftHeartbeats(t *testing.T) {
 
 	mtc := startMultiTestContext(t, 3)
 	defer mtc.Stop()
-	mtc.replicateRange(1, 0, 1, 2)
+	mtc.replicateRange(1, 1, 2)
 
 	// Capture the initial term and state.
 	status := mtc.stores[0].RaftStatus(1)
@@ -942,7 +942,7 @@ func TestReplicateAfterSplit(t *testing.T) {
 		t.Fatal(err)
 	}
 	// Now add the second replica.
-	mtc.replicateRange(rangeID2, 0, 1)
+	mtc.replicateRange(rangeID2, 1)
 
 	if mtc.stores[1].LookupReplica(roachpb.RKey(key), nil).GetMaxBytes() == 0 {
 		t.Error("Range MaxBytes is not set after snapshot applied")
@@ -1049,10 +1049,10 @@ func TestRaftAfterRemoveRange(t *testing.T) {
 	}
 
 	rangeID := roachpb.RangeID(2)
-	mtc.replicateRange(rangeID, 0, 1, 2)
+	mtc.replicateRange(rangeID, 1, 2)
 
-	mtc.unreplicateRange(rangeID, 0, 2)
-	mtc.unreplicateRange(rangeID, 0, 1)
+	mtc.unreplicateRange(rangeID, 2)
+	mtc.unreplicateRange(rangeID, 1)
 
 	// Wait for the removal to be processed.
 	util.SucceedsWithin(t, time.Second, func() error {
@@ -1088,7 +1088,7 @@ func TestRaftAfterRemoveRange(t *testing.T) {
 	}
 	// Execute another replica change to ensure that raft has processed
 	// the heartbeat just sent.
-	mtc.replicateRange(roachpb.RangeID(1), 0, 1)
+	mtc.replicateRange(roachpb.RangeID(1), 1)
 }
 
 // TestRaftRemoveRace adds and removes a replica repeatedly in an
@@ -1103,11 +1103,11 @@ func TestRaftRemoveRace(t *testing.T) {
 	defer mtc.Stop()
 
 	rangeID := roachpb.RangeID(1)
-	mtc.replicateRange(rangeID, 0, 1, 2)
+	mtc.replicateRange(rangeID, 1, 2)
 
 	for i := 0; i < 10; i++ {
-		mtc.unreplicateRange(rangeID, 0, 2)
-		mtc.replicateRange(rangeID, 0, 2)
+		mtc.unreplicateRange(rangeID, 2)
+		mtc.replicateRange(rangeID, 2)
 	}
 }
 
@@ -1124,7 +1124,7 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 
 	// Replicate the range to all stores.
 	replica := mtc.stores[0].LookupReplica(roachpb.RKeyMin, nil)
-	mtc.replicateRange(replica.RangeID, 0, 1, 2)
+	mtc.replicateRange(replica.RangeID, 1, 2)
 
 	// Initialize the gossip network.
 	var storeIDs []roachpb.StoreID
@@ -1200,7 +1200,7 @@ func TestStoreRangeRebalance(t *testing.T) {
 	store0 := mtc.stores[0]
 	replica := store0.LookupReplica(roachpb.RKeyMin, nil)
 	desc := replica.Desc()
-	mtc.replicateRange(desc.RangeID, 0, 1, 2)
+	mtc.replicateRange(desc.RangeID, 1, 2)
 
 	// Initialize the gossip network with fake capacity data.
 	storeDescs := make([]*roachpb.StoreDescriptor, 0, len(mtc.stores))
@@ -1262,7 +1262,7 @@ func TestReplicateRogueRemovedNode(t *testing.T) {
 
 	// First put the range on all three nodes.
 	raftID := roachpb.RangeID(1)
-	mtc.replicateRange(raftID, 0, 1, 2)
+	mtc.replicateRange(raftID, 1, 2)
 
 	// Put some data in the range so we'll have something to test for.
 	incArgs := incrementArgs([]byte("a"), 5)
@@ -1275,8 +1275,8 @@ func TestReplicateRogueRemovedNode(t *testing.T) {
 
 	// Stop node 2; while it is down remove the range from nodes 2 and 1.
 	mtc.stopStore(2)
-	mtc.unreplicateRange(raftID, 0, 2)
-	mtc.unreplicateRange(raftID, 0, 1)
+	mtc.unreplicateRange(raftID, 2)
+	mtc.unreplicateRange(raftID, 1)
 
 	// Make a write on node 0; this will not be replicated because 0 is the only node left.
 	incArgs = incrementArgs([]byte("a"), 11)
@@ -1355,7 +1355,7 @@ func TestReplicateReAddAfterDown(t *testing.T) {
 
 	// First put the range on all three nodes.
 	raftID := roachpb.RangeID(1)
-	mtc.replicateRange(raftID, 0, 1, 2)
+	mtc.replicateRange(raftID, 1, 2)
 
 	// Put some data in the range so we'll have something to test for.
 	incArgs := incrementArgs([]byte("a"), 5)
@@ -1369,7 +1369,7 @@ func TestReplicateReAddAfterDown(t *testing.T) {
 	// Stop node 2; while it is down remove the range from it. Since the node is
 	// down it won't see the removal and clean up its replica.
 	mtc.stopStore(2)
-	mtc.unreplicateRange(raftID, 0, 2)
+	mtc.unreplicateRange(raftID, 2)
 
 	// Perform another write.
 	incArgs = incrementArgs([]byte("a"), 11)
@@ -1386,7 +1386,7 @@ func TestReplicateReAddAfterDown(t *testing.T) {
 	// process. An ill-timed GC has been known to cause bugs including
 	// https://github.com/cockroachdb/cockroach/issues/2873.
 	mtc.restartStore(2)
-	mtc.replicateRange(raftID, 0, 2)
+	mtc.replicateRange(raftID, 2)
 
 	// The range should be synced back up.
 	mtc.waitForValues(roachpb.Key("a"), 3*time.Second, []int64{16, 16, 16})
@@ -1406,9 +1406,9 @@ func TestLeaderRemoveSelf(t *testing.T) {
 	// helps avoid a deadlock at shutdown.
 	mtc.stores[0].DisableReplicaGCQueue(true)
 	raftID := roachpb.RangeID(1)
-	mtc.replicateRange(raftID, 0, 1)
+	mtc.replicateRange(raftID, 1)
 	// Remove the replica from first store.
-	mtc.unreplicateRange(raftID, 0, 0)
+	mtc.unreplicateRange(raftID, 0)
 	getArgs := getArgs([]byte("a"))
 
 	// Force the read command request a new lease.

--- a/storage/client_replica_gc_test.go
+++ b/storage/client_replica_gc_test.go
@@ -70,8 +70,8 @@ func TestReplicaGCQueueDropReplicaDirect(t *testing.T) {
 	mtc.Start(t, numStores)
 	defer mtc.Stop()
 
-	mtc.replicateRange(rangeID, 0, 1, 2)
-	mtc.unreplicateRange(rangeID, 0, 1)
+	mtc.replicateRange(rangeID, 1, 2)
+	mtc.unreplicateRange(rangeID, 1)
 
 	// Make sure the range is removed from the store.
 	util.SucceedsWithin(t, time.Second, func() error {
@@ -93,8 +93,8 @@ func TestReplicaGCQueueDropReplicaGCOnScan(t *testing.T) {
 	mtc.stores[1].DisableReplicaGCQueue(true)
 
 	rangeID := roachpb.RangeID(1)
-	mtc.replicateRange(rangeID, 0, 1, 2)
-	mtc.unreplicateRange(rangeID, 0, 1)
+	mtc.replicateRange(rangeID, 1, 2)
+	mtc.unreplicateRange(rangeID, 1)
 
 	// Wait long enough for the direct replica GC to have had a chance and been
 	// discarded because the queue is disabled.

--- a/storage/client_replica_test.go
+++ b/storage/client_replica_test.go
@@ -55,7 +55,7 @@ func TestRangeCommandClockUpdate(t *testing.T) {
 	}
 	mtc.Start(t, numNodes)
 	defer mtc.Stop()
-	mtc.replicateRange(1, 0, 1, 2)
+	mtc.replicateRange(1, 1, 2)
 
 	// Advance the leader's clock ahead of the followers (by more than
 	// MaxOffset but less than the leader lease) and execute a command.

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -736,8 +736,8 @@ func setupSplitSnapshotRace(t *testing.T) (mtc *multiTestContext, leftKey roachp
 	leftRangeID := mtc.stores[0].LookupReplica(roachpb.RKey("a"), nil).RangeID
 
 	// Replicate the left range onto nodes 1-3 and remove it from node 0.
-	mtc.replicateRange(leftRangeID, 0, 1, 2, 3)
-	mtc.unreplicateRange(leftRangeID, 0, 0)
+	mtc.replicateRange(leftRangeID, 1, 2, 3)
+	mtc.unreplicateRange(leftRangeID, 0)
 	mtc.expireLeaderLeases()
 
 	mtc.waitForValues(leftKey, splitTimeout, []int64{0, 1, 1, 1, 0, 0})
@@ -755,12 +755,19 @@ func setupSplitSnapshotRace(t *testing.T) (mtc *multiTestContext, leftKey roachp
 
 	// Get the right range's ID. Since the split was performed on node
 	// 1, it is currently 11 and not 3 as might be expected.
-	rightRangeID := mtc.stores[1].LookupReplica(roachpb.RKey("z"), nil).RangeID
+	var rightRangeID roachpb.RangeID
+	util.SucceedsWithin(t, splitTimeout, func() error {
+		rightRangeID = mtc.stores[1].LookupReplica(roachpb.RKey("z"), nil).RangeID
+		if rightRangeID == leftRangeID {
+			return util.Errorf("store 1 hasn't processed split yet")
+		}
+		return nil
+	})
 
 	// Relocate the right range onto nodes 3-5.
-	mtc.replicateRange(rightRangeID, 1, 4, 5)
-	mtc.unreplicateRange(rightRangeID, 1, 2)
-	mtc.unreplicateRange(rightRangeID, 1, 1)
+	mtc.replicateRange(rightRangeID, 4, 5)
+	mtc.unreplicateRange(rightRangeID, 2)
+	mtc.unreplicateRange(rightRangeID, 1)
 
 	mtc.waitForValues(rightKey, splitTimeout, []int64{0, 0, 0, 2, 2, 2})
 
@@ -958,7 +965,7 @@ func TestLeaderAfterSplit(t *testing.T) {
 	mtc.Start(t, 3)
 	defer mtc.Stop()
 
-	mtc.replicateRange(1, 0, 1, 2)
+	mtc.replicateRange(1, 1, 2)
 
 	leftKey := roachpb.Key("a")
 	splitKey := roachpb.Key("m")

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -45,6 +45,7 @@ import (
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/keys"
 	"github.com/cockroachdb/cockroach/kv"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/rpc"
@@ -535,6 +536,38 @@ func (m *multiTestContext) restartStore(i int) {
 	m.senders[i].AddStore(m.stores[i])
 }
 
+// findStartKeyLocked returns the start key of the given range.
+func (m *multiTestContext) findStartKeyLocked(rangeID roachpb.RangeID) roachpb.RKey {
+	// We can use the first store that returns results because the start
+	// key never changes.
+	for _, s := range m.stores {
+		rep, err := s.GetReplica(rangeID)
+		if err == nil && rep.IsInitialized() {
+			return rep.Desc().StartKey
+		}
+	}
+	m.t.Fatalf("couldn't find range %s on any store", rangeID)
+	return nil // unreached, but the compiler can't tell.
+}
+
+// findMemberStoreLocked finds a non-stopped Store which is a member
+// of the given range.
+func (m *multiTestContext) findMemberStoreLocked(desc roachpb.RangeDescriptor) *storage.Store {
+	for _, s := range m.stores {
+		if s == nil {
+			// Store is stopped.
+			continue
+		}
+		for _, r := range desc.Replicas {
+			if s.StoreID() == r.StoreID {
+				return s
+			}
+		}
+	}
+	m.t.Fatalf("couldn't find a live member of %s", desc)
+	return nil // unreached, but the compiler can't tell.
+}
+
 // restart stops and restarts all stores but leaves the engines intact,
 // so the stores should contain the same persistent storage as before.
 func (m *multiTestContext) restart() {
@@ -547,20 +580,32 @@ func (m *multiTestContext) restart() {
 }
 
 // replicateRange replicates the given range onto the given stores.
-func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, sourceStoreIndex int, dests ...int) {
+func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, dests ...int) {
 	m.mu.RLock()
 	defer m.mu.RUnlock()
-	rng, err := m.stores[sourceStoreIndex].GetReplica(rangeID)
-	if err != nil {
-		m.t.Fatal(err)
-	}
+
+	startKey := m.findStartKeyLocked(rangeID)
 
 	for _, dest := range dests {
-		err = rng.ChangeReplicas(roachpb.ADD_REPLICA,
+		// Perform a consistent read to get the range descriptor, to make
+		// sure we have the effects of the previous ChangeReplicas call.
+		// By the time ChangeReplicas returns the raft leader is
+		// guaranteed to have the updated version, but followers are not.
+		var desc roachpb.RangeDescriptor
+		if err := m.db.GetProto(keys.RangeDescriptorKey(startKey), &desc); err != nil {
+			m.t.Fatal(err)
+		}
+
+		rep, err := m.findMemberStoreLocked(desc).GetReplica(rangeID)
+		if err != nil {
+			m.t.Fatal(err)
+		}
+
+		err = rep.ChangeReplicas(roachpb.ADD_REPLICA,
 			roachpb.ReplicaDescriptor{
 				NodeID:  m.stores[dest].Ident.NodeID,
 				StoreID: m.stores[dest].Ident.StoreID,
-			}, rng.Desc())
+			}, &desc)
 		if err != nil {
 			m.t.Fatal(err)
 		}
@@ -571,7 +616,7 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, sourceStoreIn
 		for _, dest := range dests {
 			// Use LookupRange(keys) instead of GetRange(rangeID) to ensure that the
 			// snapshot has been transferred and the descriptor initialized.
-			if m.stores[dest].LookupReplica(rng.Desc().StartKey, nil) == nil {
+			if m.stores[dest].LookupReplica(startKey, nil) == nil {
 				return util.Errorf("range not found on store %d", dest)
 			}
 		}
@@ -579,38 +624,31 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, sourceStoreIn
 	})
 }
 
-// unreplicateRange removes a replica of the range in the source store
-// from the dest store.
-func (m *multiTestContext) unreplicateRange(rangeID roachpb.RangeID, source, dest int) {
+// unreplicateRange removes a replica of the range from the dest store.
+func (m *multiTestContext) unreplicateRange(rangeID roachpb.RangeID, dest int) {
 	m.mu.RLock()
 	defer m.mu.RUnlock()
-	rng, err := m.stores[source].GetReplica(rangeID)
+
+	startKey := m.findStartKeyLocked(rangeID)
+
+	var desc roachpb.RangeDescriptor
+	if err := m.db.GetProto(keys.RangeDescriptorKey(startKey), &desc); err != nil {
+		m.t.Fatal(err)
+	}
+
+	rep, err := m.findMemberStoreLocked(desc).GetReplica(rangeID)
 	if err != nil {
 		m.t.Fatal(err)
 	}
 
-	err = rng.ChangeReplicas(roachpb.REMOVE_REPLICA,
+	err = rep.ChangeReplicas(roachpb.REMOVE_REPLICA,
 		roachpb.ReplicaDescriptor{
 			NodeID:  m.idents[dest].NodeID,
 			StoreID: m.idents[dest].StoreID,
-		}, rng.Desc())
+		}, &desc)
 	if err != nil {
 		m.t.Fatal(err)
 	}
-
-	// Wait for the source node's ReplicaDescriptor to update.
-	util.SucceedsWithin(m.t, replicationTimeout, func() error {
-		rngDesc := m.stores[source].LookupReplica(rng.Desc().StartKey, nil)
-		if rngDesc == nil {
-			return util.Errorf("range not found on store %d", source)
-		}
-		for _, replDesc := range rngDesc.Desc().Replicas {
-			if replDesc.StoreID == m.idents[dest].StoreID {
-				return util.Errorf("store %d not yet removed", source)
-			}
-		}
-		return nil
-	})
 }
 
 // readIntFromEngines reads the current integer value at the given key

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -504,11 +504,11 @@ func (r *Replica) redirectOnOrAcquireLeaderLease(trace *tracer.Trace) *roachpb.E
 	return pErr
 }
 
-// isInitialized is true if we know the metadata of this range, either
+// IsInitialized is true if we know the metadata of this range, either
 // because we created it or we have received an initial snapshot from
 // another node. It is false when a range has been created in response
 // to an incoming message but we are waiting for our initial snapshot.
-func (r *Replica) isInitialized() bool {
+func (r *Replica) IsInitialized() bool {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	return r.isInitializedLocked()
@@ -1563,7 +1563,7 @@ func (r *Replica) executeBatch(batch engine.Engine, ms *engine.MVCCStats, ba roa
 // should gossip; the bool returned indicates whether it's us.
 func (r *Replica) getLeaseForGossip(ctx context.Context) (bool, *roachpb.Error) {
 	// If no Gossip available (some tests) or range too fresh, noop.
-	if r.store.Gossip() == nil || !r.isInitialized() {
+	if r.store.Gossip() == nil || !r.IsInitialized() {
 		return false, roachpb.NewErrorf("no gossip or range not initialized")
 	}
 	var hasLease bool
@@ -1654,7 +1654,7 @@ func (r *Replica) maybeGossipFirstRange() *roachpb.Error {
 // need to avoid deadlocking in redirectOnOrAcquireLeaderLease.
 // TODO(tschottdorf): Can possibly simplify.
 func (r *Replica) maybeGossipSystemConfig() {
-	if r.store.Gossip() == nil || !r.isInitialized() {
+	if r.store.Gossip() == nil || !r.IsInitialized() {
 		return
 	}
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -1061,7 +1061,7 @@ func (s *Store) AddReplicaTest(rng *Replica) error {
 // the same Range ID has already been added to this store.
 // addReplicaInternalLocked requires that the store lock is held.
 func (s *Store) addReplicaInternalLocked(rng *Replica) error {
-	if !rng.isInitialized() {
+	if !rng.IsInitialized() {
 		return util.Errorf("attempted to add uninitialized range %s", rng)
 	}
 
@@ -1149,7 +1149,7 @@ func (s *Store) processRangeDescriptorUpdate(rng *Replica) error {
 
 // processRangeDescriptorUpdateLocked requires that the store lock is held.
 func (s *Store) processRangeDescriptorUpdateLocked(rng *Replica) error {
-	if !rng.isInitialized() {
+	if !rng.IsInitialized() {
 		return util.Errorf("attempted to process uninitialized range %s", rng)
 	}
 
@@ -1706,7 +1706,7 @@ func (s *Store) cacheReplicaDescriptorLocked(groupID roachpb.RangeID, replica ro
 // the replica).
 // canApplySnapshot requires that the store lock is held.
 func (s *Store) canApplySnapshotLocked(rangeID roachpb.RangeID, snap raftpb.Snapshot) bool {
-	if r, ok := s.mu.replicas[rangeID]; ok && r.isInitialized() {
+	if r, ok := s.mu.replicas[rangeID]; ok && r.IsInitialized() {
 		// We have the range and it's initialized, so let the snapshot
 		// through.
 		return true

